### ISSUE TITLE
v2.1.0 throws error when used with next.js

### ISSUE THREAD

**Issue by drejohnson** [styled-components/styled-components#934]
```
Error: Nesting detected at 1:289. Unfortunately nesting is not supported by styled-jsx.
    at Function.disableNestingPlugin (/home/dre/workspaces/reactjs/phreshr/node_modules/styled-jsx/dist/lib/style-transform.js:19:13)
    at proxy (/home/dre/workspaces/reactjs/phreshr/node_modules/stylis/stylis.js:1133:30)
    at compile (/home/dre/workspaces/reactjs/phreshr/node_modules/stylis/stylis.js:679:10)
    at compile (/home/dre/workspaces/reactjs/phreshr/node_modules/stylis/stylis.js:285:17)
    at stylis (/home/dre/workspaces/reactjs/phreshr/node_modules/stylis/stylis.js:1270:16)
    at stringifyRules (/home/dre/workspaces/reactjs/phreshr/node_modules/styled-components/lib/utils/stringifyRules.js:27:31)
    at ComponentStyle.generateAndInjectStyles (/home/dre/workspaces/reactjs/phreshr/node_modules/styled-components/lib/models/ComponentStyle.js:59:24)
    at StyledComponent.generateAndInjectStyles (/home/dre/workspaces/reactjs/phreshr/node_modules/styled-components/lib/models/StyledComponent.js:124:38)
    at StyledComponent.componentWillMount (/home/dre/workspaces/reactjs/phreshr/node_modules/styled-components/lib/models/StyledComponent.js:153:39)
```

**Comment by philpl** [comment]
Can you post some code for reproduction purposes and your version of next.js? It seems to be related to their code somehow, not ours, since the error includes some styled-jsx stuff

**Comment by andyngo** [comment]
Having the same issue. I'm on `2.1.0 of styled-components` and `3.0.0-beta14 of next`.

Below is an example of the code that causes the error:
`
const List = styled.div`
  border-bottom: 1px solid red;

  &:last-of-type {
    border-bottom: none;
  }
`
`

I have a custom `_document.js`, `.babelrc` as well as `package.json` configs as provided in the example here: https://github.com/zeit/next.js/tree/v3-beta/examples/with-styled-components

**Comment by andyngo** [comment]
Downgraded and tested with `2.0.1 of styled-components` and nesting still works.

**Comment by philpl** [comment]
This is a bug with styled-jsx – or at least an inter-compatibility problem. So we should raise an issue with next.js and styled-jsx.

cc @giuseppeg; any idea what this could be?

**Comment by giuseppeg** [comment]
uhm probably Stylis is shared and a singleton and therefore carries over the plugins we register https://github.com/zeit/styled-jsx/blob/master/src/lib/style-transform.js#L66-L71 ? Are you folks using styled-jsx too? Have you tried to remove `styled-jsx` from the Next.js babel preset? See https://github.com/zeit/next.js/blob/v3-beta/examples/with-styled-jsx-postcss/babel-preset.js Obviously in your case you should remove it instead of replacing, maybe use `filter` instead of `map` and then make sure that you register the custom preset https://github.com/zeit/next.js/blob/v3-beta/examples/with-styled-jsx-postcss/.babelrc

**Comment by philpl** [comment]
@giuseppeg That does sounds like the problem. Maybe we should detect whether stylis has loaded some plugins? Sounds reasonable if it's possible.

So for everyone in this thread: Please either turn off **experimental preprocessing**, or turn off the **styled-jsx** plugin.

**Comment by giuseppeg** [comment]
As a workaround you may want to remove all the plugins before using stylis `stylis.use(null)` (I will probably add that to styled-jsx too).

Probably though this should be fixed in Stylis maybe by making it a class or factory rather than a singleton cc @thysultan

**Comment by philpl** [comment]
@giuseppeg This should fix it for the babel plugin: https://github.com/styled-components/babel-plugin-styled-components/pull/62

However, for the runtime, I'm reluctant to include a reset for each time that we use stylis. It seems that a factory would fix this issue in a cleaner manner.

**Comment by thysultan** [comment]
Sure i can add this to stylis in 3.2.0, which would allow you to use it either via

```
// factory pattern
var {stylis} = require('stylis')
```

or

```
// default singleton pattern
var stylis = require('stylis')
```

**Comment by philpl** [comment]
@thysultan I'm not sure whether that's a good API :/ Sorry, but `x` vs `x.stylis` and similarly even `x.default` vs `x.stylis` is for me as random as for example `stylis.use(null)`. Maybe we can expose sth more explicit maybe?

You know what they say: Write once, maintain forever :wink: So it'd be neat to have some more explicit API

**Comment by thysultan** [comment]
@philpl

```js
var {stylis} = require('stylis')
```

Would extract a new stylis instance like you would a class `var stylis = new require('stylis')`, without adding any overhead checks to the compiler / require major sem-version bump.

What API addition would you propose?

**Comment by philpl** [comment]
@thysultan maybe you could go with a mutable API, since it's already a mutable singleton? So `.clone` or something similar, and migrate to a class pattern or factory pattern in the future (next major release?)

**Comment by thysultan** [comment]
> maybe you could go with a mutable API, since it's already a mutable singleton? So `.clone` or something similar.

@philpl That's essentially how `var {stylis} = require('stylis')` would work. You'd rather it be a method call like `var stylis = stylis.clone()`?

```js
// identical
var {stylis} = require('stylis') // uses a property getter
var stylis = require('stylis').clone() // method

// default pattern works as you'd expect
var stylis = require('stylis')
```

**Comment by philpl** [comment]
@thysultan yea, essentially, like I mentioned, I don't have any problem with the proposed API – or any other one per se – the problem is clarity. So, `stylis.stylis` doesn't really convey that it's a factory for a new instance in this case.

**Comment by thysultan** [comment]
Implementing it as `new stylis(opts)` doesn't cause any overhead when in strict mode so there that option as well.

```js
var stylis = new require('stylis')
```

**Comment by whmountains** [comment]
Please at least do something, because until then styled-components is basically unusable with next js.  🙏

**Comment by giuseppeg** [comment]
@whmountains you can do what I suggested above https://github.com/styled-components/styled-components/issues/934#issuecomment-310840909

**Comment by thysultan** [comment]
@philpl @giuseppeg The `x = new stylis` pattern landed with 3.2.0

**Comment by philpl** [comment]
@thysultan I'll push some patches for the babel plugin and SC later then :smile: Awesome work :+1:

**Comment by giuseppeg** [comment]
Fixed in `styled-jsx@1.0.9` we are using an instance of `stylis` now. This should fix the OP issue when Next upgrades to the latest v of styled-jsx.

**Comment by philpl** [comment]
@giuseppeg Yea, this has already been fixed in SC since 17 days haha. Sorry for forgetting to close this issue https://github.com/styled-components/styled-components/releases/tag/v2.1.1
